--- Kheire --- 

Work in progress. 

Data Cleaning Phase 1 is ready. 

Data Cleaning Phase 2 is work in progress. It is done on the scraped data. 

This Notebook illustrates the different steps taken to do the data cleaning and feature engineering.

In [1]:
# import necessary libraries
%load_ext autoreload
%autoreload 2
import numpy as np
import pandas as pd
from utils import *
import re

# Data Cleaning Phase 1

This the first phase of data cleaning. It is very basic and is done on the data format and columns names. 

The first thing noticed was that when reading the original csv file "books.csv" some contents of sepicific sections in specific rows were separated by commas. This caused an error, because pd.read_csv considered them as content belonging to different sections (i.e. columns), resulting in different number of columns for several rows. Luckily, these content were separate by a comma and space ", ", which made it easy to solve the problem by replacing the ", " by "/" as seen below. "books_updated.csv" is the new updated csv which is read by pandas with success.

In [3]:
### COMMENT THE FOLLOWING CELL IF YOU HAVE ALREADY RUN IT BEFORE AND SAVED THE books_updated.csv ###
# Specify the filename
filename = 'books.csv'
updated_filename = "books_updated.csv"

## In some cases in certain fields such us author the authors are separated by ", " 
## to avoid problems when reading the csv directly using pandas and to retain all rows the following steps done

# Open the file and read lines
with open(filename, 'r', encoding='utf-8') as file: # utf-8 encoding to support all languages since there are non-english content
    lines = file.readlines()

# Process the lines to handle unwanted delimiters
cleaned_lines = []
is_firstline = True # used to avoid updating the first line

for line in lines:
    
    if is_firstline: # if it is the first line i.e. Header do not update it
        is_firstline = False
        cleaned_lines.append(line)
        continue
    cleaned_line = line.replace(', ', '/')

    # Append the cleaned line
    cleaned_lines.append(cleaned_line)

# rejoin all the lines to create a new updated csv
with open(updated_filename, 'w', encoding='utf-8') as output_file:
    for cleaned_line in cleaned_lines:
        output_file.write(cleaned_line + '\n')


In [4]:
# Read the cleaned data into a DataFrame
df = pd.read_csv(updated_filename, delimiter = ",", encoding = 'utf-8', index_col=False) 

# Look at the first 5 rows of the DataFrame
df.head()

bookID                                              title  \
0      1  Harry Potter and the Half-Blood Prince (Harry ...   
1      2  Harry Potter and the Order of the Phoenix (Har...   
2      4  Harry Potter and the Chamber of Secrets (Harry...   
3      5  Harry Potter and the Prisoner of Azkaban (Harr...   
4      8  Harry Potter Boxed Set  Books 1-5 (Harry Potte...   

                      authors  average_rating        isbn         isbn13  \
0  J.K. Rowling/Mary GrandPré            4.57  0439785960  9780439785969   
1  J.K. Rowling/Mary GrandPré            4.49  0439358078  9780439358071   
2                J.K. Rowling            4.42  0439554896  9780439554893   
3  J.K. Rowling/Mary GrandPré            4.56  043965548X  9780439655484   
4  J.K. Rowling/Mary GrandPré            4.78  0439682584  9780439682589   

  language_code    num_pages  ratings_count text_reviews_count  \
0           eng          652        2095690              27591   
1           eng          870        2153167              29221   
2           eng          352           6333                244   
3           eng          435        2339585              36325   
4           eng         2690          41428                164   

  publication_date        publisher  
0        9/16/2006  Scholastic Inc.  
1         9/1/2004  Scholastic Inc.  
2        11/1/2003       Scholastic  
3         5/1/2004  Scholastic Inc.  
4        9/13/2004       Scholastic

Looking closer at the column names, we notice that in one of the column '  num_pages' the name starts with space. For ease of use, it is better to remove this unnecessary space.

In [5]:
df.columns

Index(['bookID', 'title', 'authors', 'average_rating', 'isbn', 'isbn13',
       'language_code', '  num_pages', 'ratings_count', 'text_reviews_count',
       'publication_date', 'publisher'],
      dtype='object')

In [6]:
# Remove the space before the column num_pages, for ease of use
df.rename(columns={'  num_pages': '  num_pages'.replace(' ', '')}, inplace=True)
df.columns

Index(['bookID', 'title', 'authors', 'average_rating', 'isbn', 'isbn13',
       'language_code', 'num_pages', 'ratings_count', 'text_reviews_count',
       'publication_date', 'publisher'],
      dtype='object')

Examining the column types, it is noticed that some of the columns that are supposed to be numerical types are object types

In [7]:
types_columns=df.dtypes
number_lines,number_columns=df.shape
print(types_columns)

bookID                 object
title                  object
authors                object
average_rating        float64
isbn                   object
isbn13                 object
language_code          object
num_pages               int64
ratings_count           int64
text_reviews_count     object
publication_date       object
publisher              object
dtype: object


In [8]:
# investigate if their are non-numerical non-date types in supposedly numerical and date columns

# text_reviews_count 
# Filter the DataFrame to get non-numerical values in the specified column
def check_non_numerical_date(columns_list: list, intended_type: str):
    
    if intended_type == "numerical":
        for column in columns_list:
            non_numerical_values = df.loc[~df[column].apply(pd.to_numeric, errors='coerce').notna()]
            print("The non_numerical_values in column {} : ".format(column), non_numerical_values[column])

    elif intended_type == "date":
        for column in columns_list:
            non_numerical_values = df.loc[~df[column].apply(pd.to_datetime, errors='coerce').notna()]
            print("The non_date_values in column {} : ".format(column), non_numerical_values[column])
    
# Display the non-numerical and non_date values of the selected columns
columns = ["isbn", "isbn13", "text_reviews_count"]
check_non_numerical_date(columns, "numerical")
check_non_numerical_date(["publication_date"], "date")

The non_numerical_values in column isbn :  3        043965548X
12       076790818X
16       076790382X
27       097669400X
40       006076273X
            ...    
11090    030727411X
11101    074347788X
11106    057305133X
11110    843221728X
11115    972233168X
Name: isbn, Length: 985, dtype: object
The non_numerical_values in column isbn13 :  1847    en-US
Name: isbn13, dtype: object
The non_numerical_values in column text_reviews_count :  1847    9/1/2003
Name: text_reviews_count, dtype: object
The non_date_values in column publication_date :  1847      MTV Books
8180     11/31/2000
11098     6/31/1982
Name: publication_date, dtype: object


From the above investigation one can notice the following : 
- isbn is not numerical since some of the isbn codes end with X 
- we notice that at index location 1847 the data is shifted to the left
- dates at index locations 8180 and 11098 were not considered as dates since they are wrong. June and November are 31 days, but in reality they are only 30 days.

Also, as seen below there is one empty cell in publisher column, which is on index location 1847.

In [9]:
df.isna().sum()

bookID                0
title                 0
authors               0
average_rating        0
isbn                  0
isbn13                0
language_code         0
num_pages             0
ratings_count         0
text_reviews_count    0
publication_date      0
publisher             1
dtype: int64

In [10]:
df["publisher"][df["publisher"].isna()==True]

1847    NaN
Name: publisher, dtype: object

In [11]:
df.iloc[1847]

bookID                6549/ said the shotgun to the head.
title                                       Saul Williams
authors                                              4.22
average_rating                                743470796.0
isbn                                        9780743470797
isbn13                                              en-US
language_code                                         192
num_pages                                            2762
ratings_count                                         214
text_reviews_count                               9/1/2003
publication_date                                MTV Books
publisher                                             NaN
Name: 1847, dtype: object

One can notice that there is a / in the bookID which separates the BookID from the book title. This must be a result from updating automatically the csv files at the beginning. 

Below is the code to fix this unwanted mistake

In [12]:
bookid_title = df["bookID"].iloc[1847].split('/')


# content of the remaining columns
remaining_columns = df[['title', 'authors', 'average_rating', 'isbn', 'isbn13',
                                   'language_code', 'num_pages', 'ratings_count', 
                                   'text_reviews_count', 'publication_date']].iloc[1847]


# rearrange cells content for each column
df.iloc[1847, 2:] = remaining_columns
df["bookID"].iloc[1847]= bookid_title[0]
df["title"].iloc[1847]= bookid_title[1]

C:\Users\cleme\AppData\Local\Temp\ipykernel_26336\3963356864.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["bookID"].iloc[1847]= bookid_title[0]
C:\Users\cleme\AppData\Local\Temp\ipykernel_26336\3963356864.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["title"].iloc[1847]= bookid_title[1]


In [13]:
# rechek the content 
df.iloc[1847]

bookID                                          6549
title                  said the shotgun to the head.
authors                                Saul Williams
average_rating                                  4.22
isbn                                     743470796.0
isbn13                                 9780743470797
language_code                                  en-US
num_pages                                        192
ratings_count                                   2762
text_reviews_count                               214
publication_date                            9/1/2003
publisher                                  MTV Books
Name: 1847, dtype: object

In [14]:
# Check the types of the columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11127 entries, 0 to 11126
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   bookID              11127 non-null  object
 1   title               11127 non-null  object
 2   authors             11127 non-null  object
 3   average_rating      11127 non-null  object
 4   isbn                11127 non-null  object
 5   isbn13              11127 non-null  object
 6   language_code       11127 non-null  object
 7   num_pages           11127 non-null  object
 8   ratings_count       11127 non-null  int64 
 9   text_reviews_count  11127 non-null  object
 10  publication_date    11127 non-null  object
 11  publisher           11127 non-null  object
dtypes: int64(1), object(11)
memory usage: 1.0+ MB


For some reasons the supposedly numerical columns became objects 
Re-investigate again if they contain non-numerical content

In [15]:
# re-investigate if there is non-numerical content in supposedly numerical columns
columns = ["average_rating", "isbn13", "text_reviews_count", "num_pages", "ratings_count", "text_reviews_count"]
check_non_numerical_date(columns, "numerical")

The non_numerical_values in column average_rating :  Series([], Name: average_rating, dtype: object)
The non_numerical_values in column isbn13 :  Series([], Name: isbn13, dtype: object)
The non_numerical_values in column text_reviews_count :  Series([], Name: text_reviews_count, dtype: object)
The non_numerical_values in column num_pages :  Series([], Name: num_pages, dtype: object)
The non_numerical_values in column ratings_count :  Series([], Name: ratings_count, dtype: int64)
The non_numerical_values in column text_reviews_count :  Series([], Name: text_reviews_count, dtype: object)


They do not contain non-numerical content, so will transform them to numerical

In [16]:
# convert to numerical
def convert_to_numerical(columns_list: list):
    
    for column in columns_list:
        df[column] = df[column].apply(pd.to_numeric)

columns = ["average_rating", "isbn13", "text_reviews_count", "num_pages", "ratings_count", "text_reviews_count"]

convert_to_numerical(columns)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11127 entries, 0 to 11126
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   bookID              11127 non-null  object 
 1   title               11127 non-null  object 
 2   authors             11127 non-null  object 
 3   average_rating      11127 non-null  float64
 4   isbn                11127 non-null  object 
 5   isbn13              11127 non-null  int64  
 6   language_code       11127 non-null  object 
 7   num_pages           11127 non-null  int64  
 8   ratings_count       11127 non-null  int64  
 9   text_reviews_count  11127 non-null  int64  
 10  publication_date    11127 non-null  object 
 11  publisher           11127 non-null  object 
dtypes: float64(1), int64(4), object(7)
memory usage: 1.0+ MB


In [17]:
# fix the dates by replacing 31 with 30 in June and November
# convert the date column to type pd.datetime
df["publication_date"].iloc[8180] = '11/30/2000'
df["publication_date"].iloc[11098] = '6/30/1982'

C:\Users\cleme\AppData\Local\Temp\ipykernel_26336\3398859105.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["publication_date"].iloc[8180] = '11/30/2000'
C:\Users\cleme\AppData\Local\Temp\ipykernel_26336\3398859105.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["publication_date"].iloc[11098] = '6/30/1982'


Resave the DataFrame to avoid repeating the process each time we want to use the df

In [18]:
df.to_csv("books_updated.csv", index_label=False) # index_label = False so that it does not add another index label to the DataFrame

# Data Cleaning Phase 2

This data cleaning is done on the data set after doing webscraping. Through webscraping extra features were extracted which are: 

- first_publish : this is the date a book was first published (this will help distinguish two different books with the same name)
- book_format : this is the format of the book (some books are present in different formats (paperback, Audio CD, hardcover))
- new_publisher: this is because it was noticed that in the original data some books had wrong publishers
- edition_avgRating: this is the actaul average rating of each edition
- added_toShelves: this is the number of users that added a book to shelves

*For more details about the process of scraping, please refer to scraper/scraper.py*

### Read the Data

In [463]:
# read the scraped data
df_scraped = pd.read_csv("scraper/booksRating_extraFeats.csv")

In [464]:
df_scraped.head()

bookID                                              title  \
0       1  Harry Potter and the Half-Blood Prince (Harry ...   
1       2  Harry Potter and the Order of the Phoenix (Har...   
2       4  Harry Potter and the Chamber of Secrets (Harry...   
3       5  Harry Potter and the Prisoner of Azkaban (Harr...   
4       8  Harry Potter Boxed Set  Books 1-5 (Harry Potte...   

                      authors  average_rating        isbn         isbn13  \
0  J.K. Rowling/Mary GrandPré            4.57  0439785960  9780439785969   
1  J.K. Rowling/Mary GrandPré            4.49  0439358078  9780439358071   
2                J.K. Rowling            4.42  0439554896  9780439554893   
3  J.K. Rowling/Mary GrandPré            4.56  043965548X  9780439655484   
4  J.K. Rowling/Mary GrandPré            4.78  0439682584  9780439682589   

  language_code  num_pages  ratings_count  text_reviews_count  \
0           eng        652        2095690               27591   
1           eng        870        2153167               29221   
2           eng        352           6333                 244   
3           eng        435        2339585               36325   
4           eng       2690          41428                 164   

  publication_date        publisher  first_published            book_format  \
0       2006-09-16  Scholastic Inc.    July 16, 2005              Paperback   
1       2004-09-01  Scholastic Inc.    June 21, 2003              Paperback   
2       2003-11-01       Scholastic     July 2, 1998              Hardcover   
3       2004-05-01  Scholastic Inc.     July 8, 1999  Mass Market Paperback   
4       2004-09-13       Scholastic  October 1, 2003              Paperback   

     new_publisher  edition_avgRating  added_toShelves  
0   Scholastic Inc               4.57        4405980.0  
1  Scholastic Inc.               4.50        4518536.0  
2              NaN               4.05           7469.0  
3  Scholastic Inc.               4.57        5223956.0  
4       Scholastic               4.72         172736.0

In [465]:
df_scraped.tail()

bookID                                             title  \
11122   45631  Expelled from Eden: A William T. Vollmann Reader   
11123   45633                       You Bright and Risen Angels   
11124   45634                   The Ice-Shirt (Seven Dreams #1)   
11125   45639                                       Poor People   
11126   45641                       Las aventuras de Tom Sawyer   

                                                 authors  average_rating  \
11122  William T. Vollmann/Larry McCaffery/Michael He...            4.06   
11123                                William T. Vollmann            4.08   
11124                                William T. Vollmann            3.96   
11125                                William T. Vollmann            3.72   
11126                                         Mark Twain            3.91   

             isbn         isbn13 language_code  num_pages  ratings_count  \
11122  1560254416  9781560254416           eng        512            156   
11123  0140110879  9780140110876           eng        635            783   
11124  0140131965  9780140131963           eng        415            820   
11125  0060878827  9780060878825           eng        434            769   
11126  8497646983  9788497646987           spa        272            113   

       text_reviews_count publication_date      publisher   first_published  \
11122                  20       2004-12-21  Da Capo Press  December 1, 2004   
11123                  56       1988-12-01  Penguin Books     June 27, 1987   
11124                  95       1993-08-01  Penguin Books   January 1, 1990   
11125                 139       2007-02-27           Ecco   January 1, 2007   
11126                  12       2006-05-28  Edimat Libros      June 1, 1876   

      book_format             new_publisher  edition_avgRating  \
11122   Paperback             Da Capo Press               4.06   
11123   Paperback             Penguin Books               4.04   
11124   Paperback  Penguin Publishing Group               3.98   
11125   Hardcover                      Ecco               3.75   
11126   Paperback                       NaN               3.72   

       added_toShelves  
11122            552.0  
11123           4269.0  
11124           4686.0  
11125           2948.0  
11126            228.0

In [466]:
df_scraped.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11127 entries, 0 to 11126
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   bookID              11127 non-null  int64  
 1   title               11127 non-null  object 
 2   authors             11127 non-null  object 
 3   average_rating      11127 non-null  float64
 4   isbn                11127 non-null  object 
 5   isbn13              11127 non-null  int64  
 6   language_code       11127 non-null  object 
 7   num_pages           11127 non-null  int64  
 8   ratings_count       11127 non-null  int64  
 9   text_reviews_count  11127 non-null  int64  
 10  publication_date    11127 non-null  object 
 11  publisher           11127 non-null  object 
 12  first_published     11125 non-null  object 
 13  book_format         11126 non-null  object 
 14  new_publisher       8360 non-null   object 
 15  edition_avgRating   11124 non-null  float64
 16  adde

### Fill Missing Data

In the new dataset, there is 2 missing data in the first_published, 1 missing value in the book_format, 3 missing values in the edition_avgRating and added_toShelves. These were values that the scraper failed to get. Since they are very few, they will be filled manually. 

In [467]:
df_scraped[df_scraped[["first_published", "book_format", "edition_avgRating", "added_toShelves"]].isna().any(axis=1)]

bookID                                              title  \
2255    8077  Animales No Se Visten  Los (Animals Should Def...   
6396   24062                            The Deep (Dive Trilogy)   
8476   32552                          Essential Tales and Poems   
8520   32703   The Diary of Ellen Rimbauer: My Life at Rose Red   

                                           authors  average_rating  \
2255                      Judi Barrett/Ron Barrett            4.11   
6396                                 Gordon Korman            3.78   
8476            Edgar Allan Poe/Benjamin F. Fisher            4.36   
8520  Joyce Reardon/Steven Rimbauer/Ridley Pearson            3.67   

            isbn         isbn13 language_code  num_pages  ratings_count  \
2255  1595191356  9781595191359           eng         32              0   
6396  0613674839  9780613674836           eng        148              0   
8476  1593080646  9781593080648         en-US        688          66382   
8520  0786890436  9780786890439           eng        277           7852   

      text_reviews_count publication_date               publisher  \
2255                   0       1991-10-30          Live Oak Media   
6396                   0       2003-07-01        Turtleback Books   
8476                 109       2004-10-25  Barnes  Noble Classics   
8520                 352       2001-04-29                Hyperion   

      first_published            book_format  \
2255  January 1, 1970              Paperback   
6396              NaN                    NaN   
8476              NaN              Paperback   
8520  January 1, 2001  Mass Market Paperback   

                         new_publisher  edition_avgRating  added_toShelves  
2255  Atheneum Books for Young Readers                NaN              NaN  
6396                               NaN                NaN              NaN  
8476           Barnes & Noble Classics               4.36          92622.0  
8520                          Hyperion                NaN              NaN

In [468]:
df_scraped.loc[2255, ["edition_avgRating", "added_toShelve"]] = [4.11, 5402]
# df_scraped.loc[6396, ["first_published", "book_format", "new_publisher", "edition_avgRating", "added_toShelve"]] ### !!!! Couldn'y find exact book edition online !!!!
df_scraped.loc[8476, ["first_published"]] = ["January 1, 1843"]
df_scraped.loc[8520, ["edition_avgRating", "added_toShelve"]] = [3.7, 15942]

One of the books at index 6396 was not found on good reads even when searching by book ID, isbn and isbn13. Therefore, the edition_avgRating will be the same as the average_rating. added_toShelves will be equal to the mean of the added_toShelves of the other editions of the book; if other editions are not found the added_toShelves will be equal to the mode of the books having same ratings_count and text_reviews_count. The first_published will be the same as the publication_date and the book_format will be the mode of the books of approximately the same size. 

In [469]:
df_scraped[df_scraped['title'] == df_scraped.loc[6396, 'title']]

bookID                    title        authors  average_rating  \
6396   24062  The Deep (Dive Trilogy)  Gordon Korman            3.78   

            isbn         isbn13 language_code  num_pages  ratings_count  \
6396  0613674839  9780613674836           eng        148              0   

      text_reviews_count publication_date         publisher first_published  \
6396                   0       2003-07-01  Turtleback Books             NaN   

     book_format new_publisher  edition_avgRating  added_toShelves  \
6396         NaN           NaN                NaN              NaN   

      added_toShelve  
6396             NaN

In [470]:
df_scraped.loc[df_scraped['title'].str.contains('dive', case=False)]

bookID                                              title  \
2021    7271  Marine Conservation Biology: The Science of Ma...   
4458   16059                       The Dive From Clausen's Pier   
4459   16063                       The Dive from Clausen's Pier   
6385   24037                                The Deep (Dive  #2)   
6386   24040                           The Discovery (Dive  #1)   
6387   24043  New York City's Best Dive Bars: Drinking and D...   
6396   24062                            The Deep (Dive Trilogy)   

                                                authors  average_rating  \
2021  Elliott A. Norse/Larry B. Crowder/Michael E. S...            4.59   
4458                                         Ann Packer            3.42   
4459                                         Ann Packer            3.42   
6385                                      Gordon Korman            3.78   
6386                                      Gordon Korman            3.72   
6387                           Wendy  Mitchell/June Kim            3.75   
6396                                      Gordon Korman            3.78   

            isbn         isbn13 language_code  num_pages  ratings_count  \
2021  1559636629  9781559636629           eng        496             16   
4458  0375727132  9780375727139           eng        432          19844   
4459  0749933631  9780749933630           eng        368             60   
6385  0439507235  9780439507233           eng        148           1436   
6386  0439507227  9780439507226           eng        141           1858   
6387  0970312539  9780970312532           eng        160             16   
6396  0613674839  9780613674836           eng        148              0   

      text_reviews_count publication_date                      publisher  \
2021                   2       2005-05-09                   Island Press   
4458                1747       2003-04-08                        Vintage   
4459                   8       2003-03-27                        Piatkus   
6385                  55       2003-07-01                     Scholastic   
6386                 135       2003-06-01  Apple Paperbacks (Scholastic)   
6387                   2       2003-07-01                  Gamble Guides   
6396                   0       2003-07-01               Turtleback Books   

        first_published book_format  new_publisher  edition_avgRating  \
2021        May 9, 2005   Paperback   Island Press               4.50   
4458    January 1, 2002   Paperback        Vintage               3.44   
4459    January 1, 2002   Paperback            NaN               3.15   
6385       July 1, 2003   Paperback     Scholastic               3.78   
6386  September 1, 2005   Paperback     Scholastic               3.71   
6387   December 1, 2002   Paperback  Gamble Guides               3.71   
6396                NaN         NaN            NaN                NaN   

      added_toShelves  added_toShelve  
2021            175.0             NaN  
4458          34745.0             NaN  
4459             90.0             NaN  
6385           2837.0             NaN  
6386           3896.0             NaN  
6387             32.0             NaN  
6396              NaN             NaN

-- Clemence --

I found the book on GoodReads : https://www.goodreads.com/book/show/24037.The_Deep

As the information on the goodreads page are not exactly the same as the ones in line 6385, I filled the information manually based on the GoddReads page (the title, num_pages and average rating where the same in the database and the link to goodread).

In [471]:
df_scraped.loc[6396, ["ratings_count", "text_reviews_coun", "first_published","book_format","new_publisher","edition_avgRating","added_toShelves"]] = [1659,66,"July 1, 2003","Paperback","Scholastic",3.78,3004]

In [472]:
df_scraped.iloc[6396]

bookID                                  24062
title                 The Deep (Dive Trilogy)
authors                         Gordon Korman
average_rating                           3.78
isbn                               0613674839
isbn13                          9780613674836
language_code                             eng
num_pages                                 148
ratings_count                            1659
text_reviews_count                          0
publication_date                   2003-07-01
publisher                    Turtleback Books
first_published                  July 1, 2003
book_format                         Paperback
new_publisher                      Scholastic
edition_avgRating                        3.78
added_toShelves                        3004.0
added_toShelve                            NaN
text_reviews_coun                        66.0
Name: 6396, dtype: object

For the new_publisher, the scraper failed to get a lot of values, this is because of how the publisher is saved in the html. Due to constraint of time, and since there is already a publisher column in the original dataset. In the cases where there are NaN the old publisher will be put. The new publisher will be taken, only in the cases where the new publisher is different from the old publisher in value.

In [473]:
# Let's create the function for it. This function will only be use when the text format will be editing, 
#in the subtitle "publisher"

def update_publishers(df):
    # Iterate over each row in the DataFrame
    for index, row in df.iterrows():
        # Check if the new publisher is different from the old publisher
        if pd.notna(row['new_publisher']) and row['new_publisher'] != row['publisher']:
            # Update the publisher with the new publisher value
            df.at[index, 'publisher'] = row['new_publisher']
        else:
            # Use the old publisher if the new publisher is NaN or the same as the old publisher
            df.at[index, 'publisher'] = row['publisher']
    return df

### Set text columns to lower text

When dealing with text data it always safer to have them all in the same case either lower or upper 

In [474]:
def to_lower(df: pd.DataFrame, columns: list):
    """Function to convert text columns to lowercase"""
    for col in columns:
        if df[col].dtype == 'object':
            df[col] = df[col].str.lower()
    return df

In [475]:
# get the text columns
text_columns = df_scraped.select_dtypes(include=['object']).columns

# Convert to lowercase
df_scraped = to_lower(df_scraped, text_columns)

In [476]:
df_scraped.sample(5)

bookID                                title  \
2609    9585           halbschlaf im froschpyjama   
4009   14249                      prodigal summer   
6556   24729  after the plague: and other stories   
3666   13254                        name dropping   
2946   10910                                mercy   

                                 authors  average_rating        isbn  \
2609  tom robbins/pociao/walter hartmann            3.74  3499224429   
4009                  barbara kingsolver            4.00  0060959037   
6556                t. coraghessan boyle            3.86  0142001414   
3666                         jane heller            3.52  0312978332   
2946                        jodi picoult            3.58  0743422449   

             isbn13 language_code  num_pages  ratings_count  \
2609  9783499224423           ger        459             22   
4009  9780060959036           eng        444          88241   
6556  9780142001417           eng        303           1253   
3666  9780312978334           eng        323            701   
2946  9780743422444           eng        400          42046   

      text_reviews_count publication_date                publisher  \
2609                   2       1998-11-02           rowohlt verlag   
4009                5652       2001-10-16         harper perennial   
6556                 108       2002-12-31            penguin books   
3666                  53       2001-03-15       st. martin's press   
2946                2564       2001-04-01  washington square press   

         first_published            book_format              new_publisher  \
2609     january 1, 1994              paperback                        NaN   
4009    october 17, 2000              paperback           harper perennial   
6556  september 10, 2001              paperback   penguin publishing group   
3666      april 28, 2000  mass market paperback    st. martin's paperbacks   
2946     january 1, 1996              paperback  atria/emily bestler books   

      edition_avgRating  added_toShelves  added_toShelve  text_reviews_coun  
2609               3.42             40.0             NaN                NaN  
4009               4.02         186478.0             NaN                NaN  
6556               3.88           2554.0             NaN                NaN  
3666               3.51           1450.0             NaN                NaN  
2946               3.58          96944.0             NaN                NaN

### Take First Author Name and Create num_contributors Column

We have decided to take only the name of the first author, since usually first authors are the main authors. Another column will be added to indicate the total number of authors/contributors to the book.

In [477]:
# take the first author in the authors column
df_scraped["first_author"] = df_scraped["authors"].apply(lambda x: x.split("/")[0].strip())


In [478]:
df_scraped["first_author"].sample(5)

10704         sara gruen
5883      yvon chouinard
6551       sue monk kidd
4985      p.g. wodehouse
2747     margaret hodges
Name: first_author, dtype: object

In [479]:
df_scraped["num_contributors"] = df_scraped["authors"].apply(lambda x: len(x.split("/")))

In [480]:
df_scraped[["authors","num_contributors"]].sample(5)

authors  num_contributors
1213                   amy sedaris                 1
10496                  roddy doyle                 1
1995   cesare beccaria/david young                 2
10685                   t.h. white                 1
4684           f. scott fitzgerald                 1

### Clean Text Columns

Remove punctuation, and extra white spaces from all text columns.

Better clean the text in publisher column, to avoid having same publishers represented differently. 
Example: indexes 5993, 5365 and 9753, W.W. Norton & Company, W. W. Norton and Company and W. W. Norton  Company represent the same publisher

In [481]:
general_replacements = [
    (r'[^\w\s]', ''), # remove punctuation
    (r'\s\s+', " "), # remove double spaces and more   
]

columns_ = ["title", "first_author"]

df_scraped = sub_text(df_scraped, columns_, general_replacements)

In [482]:
df_scraped[["title", "first_author"]].sample(5)

title    first_author
2984  eight men out the black sox and the 1919 world...    eliot asinof
7160                 waterfront a walk around manhattan  phillip lopate
211                                        tesla papers    nikola tesla
9665           vegas bites a werewolf romance anthology        la banks
194   pyramids of montauk explorations in consciousness      peter moon

In [483]:
specific_replacements = [
    ("inc.", ""),
    ("llc", ""),
    ("ltd", ""),
    ("w. w.", "ww"), 
    ("&", " and "),
    (r'\bbooks?\b', ""),  
    (r'\bclassics?\b', ""),
    (r'\bpublishers?\b', ""),
    (r'\bpress\b', ""),
    (r'\bpublishing\b', "")
]

df_scraped = sub_text(df_scraped, ["publisher"], specific_replacements)

### Update publisher column with "new_publisher"

In [484]:
#Use update_publishers function to be sure about our "publisher" column : 

In [485]:
df_scraped[['publisher', 'new_publisher']].tail(50)

publisher                     new_publisher
11077                          debols!llo                               NaN
11078                             montena                               NaN
11079                             montena                               NaN
11080                   listening library                               NaN
11081           listening library (audio)         listening library (audio)
11082  alfred a. knopf  for young readers                               NaN
11083  alfred a. knopf  for young readers                               NaN
11084                   listening library                               NaN
11085                     alfred a. knopf                               NaN
11086                         ediciones b                               NaN
11087            knopf  for young readers                               NaN
11088                         scholastic                                NaN
11089                                rayo                               NaN
11090                          debolsillo                        debolsillo
11091                                rayo                               NaN
11092                            profile                                NaN
11093                     harper business                   harper business
11094                             rodale                          rodale pr
11095           william morrow paperbacks         william morrow paperbacks
11096           vintage/random house (ny)         vintage/random house (ny)
11097                             penguin                               NaN
11098                      folio histoire                               NaN
11099                          penguin  .                               NaN
11100                            penguin                                NaN
11101                simon  and  schuster                  simon & schuster
11102                             on tape                               NaN
11103                         touchstone                                NaN
11104                           not avail  alpine fine arts collection (uk)
11105                  alfaguara infantil                               NaN
11106                      samuel french                  samuel french ltd
11107                      bbc audiobooks                               NaN
11108                manning publications                           manning
11109                    punto de lectura                  punto de lectura
11110                            planeta                 planeta publishing
11111                                suma                               NaN
11112                         seix barral                               NaN
11113              panamericana editorial                               NaN
11114               harpercollins espanol                               NaN
11115                  editorial presença                               NaN
11116                  editorial presença                               NaN
11117                  editorial presença                               NaN
11118                  editorial presença                               NaN
11119                  editorial presença                               NaN
11120                  editorial presença                               NaN
11121                            penguin           penguin publishing group
11122                            da capo                      da capo press
11123                            penguin                      penguin books
11124                            penguin           penguin publishing group
11125                                ecco                              ecco
11126                       edimat libros                               NaN

-- Clemence --

Based on the sample, i'm not sure we should always take the new_publisher column... For example for line 11121, 11123 and 11124, with the "publisher" column we have the same publisher : Penguin books. With the new_publisher column we will have 2 different publishers (Penguin Publishing Group and Penbguin Books). I don't remember why you wanted to add new information about publisher ? There was trouble with the publisher column ? Otherwise I propose to work only on the publisher column if it's okay with you :) I did this below

### Add "size_of_publisher" column

In [486]:
# Adding the new columns that allows to see if a publisher is often cited
df_scraped['publisher_count'] = df_scraped.groupby('publisher')['publisher'].transform('count')

In [487]:
# The exact number of time a publisher name is in the database is not as relevant as the number of time an author or a title is
# in it.
# To simplify this information, we can, instead of the count, create a column that categorize the publisher :
# - Small publisher(1) = only 1 time in the database
# - Medium publisher(2) = from 2 to 10 times
# - big publisher(3) = more that 10 times

# Create a function to determine the size of the publisher based on the number of times it is mentioned
def determine_size(publisher_count):
    if publisher_count == 1:
        return 1
    elif publisher_count < 10:
        return 2
    else:
        return 3

# Apply the function to the 'publisher_count' column to create the new 'size_of_publisher' column
df_scraped['size_of_publisher'] = df_scraped['publisher_count'].apply(lambda x: determine_size(x))

In [488]:
print('Number of line with small publisher',len(df_scraped[df_scraped['size_of_publisher'] == 1]))
print('Number of line with medium publisher',len(df_scraped[df_scraped['size_of_publisher'] == 2]))
print('Number of line with big publisher',len(df_scraped[df_scraped['size_of_publisher'] == 3]))

Number of line with small publisher 1254
Number of line with medium publisher 2647
Number of line with big publisher 7226


In [489]:
#Now that we have the "size_of_publisher" column, let's delete the "publisher_count" used to create it 
df_scraped.drop(columns=['publisher_count'], inplace=True)

### Create "num_book_per_author" column

In [490]:
# Check for duplicates in a specific column using value_counts
num_book_per_author = df_scraped["first_author"].value_counts()

# Display values with count greater than 1 (indicating duplicates)
print(num_book_per_author[num_book_per_author > 1])

william shakespeare    88
stephen king           82
jrr tolkien            51
pg wodehouse           46
agatha christie        45
                       ..
paul farmer             2
amy sedaris             2
ruby ann boxcar         2
anthony loyd            2
james lee burke         2
Name: first_author, Length: 1471, dtype: int64


In [491]:
# Adding the new columns that allows to see if an author is often cited
df_scraped['num_book_per_author'] = df_scraped.groupby('first_author')['first_author'].transform('count')

In [492]:
df_scraped.sample(10)

bookID                                              title  \
4419    15888                      an instance of the fingerpost   
10845   44254                             la telaraña de carlota   
4418    15887   la historia de la familia roccamatio de helsinki   
8868    34298                                           dublin 4   
8184    31386                                ariel sharon a life   
7014    26590          the final battle tales from the odyssey 6   
2471     9006  dont get too comfortable the indignities of co...   
8959    34762      legends lies cherished myths of world history   
4852    17473                         black beauty coloring book   
3642    13149                         the thomas berryman number   

                                          authors  average_rating        isbn  \
4419                                   iain pears            3.94  1573227951   
10845  e.b. white/guillermo solana/garth williams            4.17  006075740x   
4418                 yann martel/bianca southwood            3.48  8423338665   
8868                                 maeve binchy            3.64  0099498588   
8184          nir hefez/gadi bloom/mitch ginsburg            3.72  1400065879   
7014                mary pope osborne/troy howell            4.08  0786809949   
2471                                 david rakoff            3.72  0767916034   
8959                             richard shenkman            3.47  0060922559   
4852               anna sewell/john         green            3.96  048629272x   
3642                              james patterson            2.86  0446600458   

              isbn13 language_code  num_pages  ratings_count  \
4419   9781573227957           eng        691          19134   
10845  9780060757403           spa        224            225   
4418   9788423338665           spa         91             14   
8868   9780099498582         en-us        256           2155   
8184   9781400065875         en-us        490             40   
7014   9780786809943         en-us        112            263   
2471   9780767916035           eng        222           8410   
8959   9780060922559           eng        320            222   
4852   9780486292724           eng         48             68   
3642   9780446600453           eng        272           4320   

       text_reviews_count  ... book_format              new_publisher  \
4419                 1081  ...   paperback   penguin publishing group   
10845                  38  ...   paperback                        NaN   
4418                    2  ...   hardcover                        NaN   
8868                   59  ...   paperback                      arrow   
8184                    6  ...   hardcover         random house, inc.   
7014                   17  ...   paperback                   hyperion   
2471                  751  ...   paperback                     anchor   
8959                   16  ...   paperback  william morrow paperbacks   
4852                    3  ...   paperback         dover publications   
3642                  307  ...   paperback   grand central publishing   

      edition_avgRating added_toShelves added_toShelve  text_reviews_coun  \
4419               3.93         46014.0            NaN                NaN   
10845              4.18          1041.0            NaN                NaN   
4418               3.05            38.0            NaN                NaN   
8868               3.65          5540.0            NaN                NaN   
8184               3.73           112.0            NaN                NaN   
7014               4.12           712.0            NaN                NaN   
2471               3.73         21973.0            NaN                NaN   
8959               3.46           520.0            NaN                NaN   
4852               4.13           151.0            NaN                NaN   
3642               2.85         11481.0            NaN                NaN   

         

### Create "is_english" column

In [493]:
# Add a new column 'is_english' with 1 for English (en, en-CA, en-GB,...) and 0 for non-English
df_scraped['is_english'] = np.where(df_scraped['language_code'].str.contains('en', case=False), 1, 0)

In [494]:
df_scraped.sample(10)

bookID                                              title  \
8198    31466                         far from the madding crowd   
6330    23846                            mischief harrow house 2   
494      1559                                   oedipus the king   
11034   45280                    quicksilver the baroque cycle 1   
9129    35515  the inner reaches of outer space metaphor as m...   
4284    15407  the lord of the rings millennium edition boxed...   
9433    37420              the short stories of ernest hemingway   
9300    36569                                     ice blue ice 3   
6543    24671                               the tristan betrayal   
4275    15351  the return of the shadow the history of the lo...   

                                               authors  average_rating  \
8198                                      thomas hardy            3.94   
6330                                     douglas clegg            3.49   
494    sophocles/bernard knox/cynthia brantley johnson            3.70   
11034                                  neal stephenson            3.93   
9129                                   joseph campbell            4.22   
4284                                    j.r.r. tolkien            4.50   
9433                                  ernest hemingway            4.26   
9300                                       anne stuart            3.94   
6543                                     robert ludlum            3.84   
4275                j.r.r. tolkien/christopher tolkien            4.03   

             isbn         isbn13 language_code  num_pages  ratings_count  \
8198   0140620478  9780140620474         en-gb        374            320   
6330   0843947667  9780843947663           eng        359            385   
494    1416500332  9781416500339           eng        144            740   
11034  0434008176  9780434008179           eng        927            171   
9129   1577312090  9781577312093         en-gb        160            971   
4284   0618037667  9780618037667           eng       1472            120   
9433   0020518609  9780020518600           eng        499            113   
9300   0778324788  9780778324782           eng        378           3215   
6543   0312990685  9780312990688           eng        505           5942   
4275   061808357x  9780618083572           eng        512           2343   

       text_reviews_count  ...            new_publisher edition_avgRating  \
8198                   42  ...                      NaN              3.77   
6330                   29  ...            leisure books              3.40   
494                    57  ...                      NaN              3.62   
11034                  12  ...                      NaN              4.08   
9129                   49  ...        new world library              4.22   
4284                    5  ...                      NaN              4.80   
9433                   12  ...                      NaN              3.89   
9300                  197  ...                     mira              3.93   
6543                  150  ...  st. martin's paperbacks              3.86   
4275                   92  ...         houghton mifflin              4.07   

      added_toShelves added_toShelve text_reviews_coun      first_author  \
8198            917.0            NaN               NaN      thomas hardy   
6330           1008.0            NaN               NaN     douglas clegg   
494            1588.0            NaN               NaN         sophocles   
11034           355.0            NaN               NaN   neal stephenson   
9129           3512.0            NaN               NaN   joseph campbell   
4284            192.0            NaN               NaN       jrr tolkien   
9433            422.0            NaN               NaN  ernest hemingway   
9300           6017.0            NaN               NaN       anne stuart   
6543          10694.0            NaN               NaN     robert ludlum  

### Create "book_count" column

In [495]:
# Adding the new columns that allows to see if a book is often cited
df_scraped['book_count'] = df_scraped.groupby('title')['title'].transform('count')

In [496]:
df_scraped.query("title == 'the brothers karamazov'")

bookID                   title  \
1406    4933  the brothers karamazov   
1407    4934  the brothers karamazov   
1408    4935  the brothers karamazov   
1409    4936  the brothers karamazov   
1410    4938  the brothers karamazov   
1411    4940  the brothers karamazov   
1634    5691  the brothers karamazov   
1990    7135  the brothers karamazov   
9356   37058  the brothers karamazov   

                                                authors  average_rating  \
1406  fyodor dostoyevsky/constance garnett/manuel ko...            4.32   
1407  fyodor dostoyevsky/fyodor dostoevsky/richard p...            4.32   
1408                    fyodor dostoyevsky/david mcduff            4.32   
1409  fyodor dostoyevsky/richard pevear/larissa volo...            4.32   
1410  fyodor dostoyevsky/simon vance/thomas r. beyer...            4.32   
1411  fyodor dostoyevsky/constance garnett/maire jaanus            4.32   
1634  fyodor dostoyevsky/richard pevear/larissa volo...            4.32   
1990  fyodor dostoyevsky/andrew r. macandrew/konstan...            4.32   
9356  fyodor dostoyevsky/thomas r. beyer jr./simon v...            4.32   

            isbn         isbn13 language_code  num_pages  ratings_count  \
1406  0451527348  9780451527349           eng        736            983   
1407  0374528373  9780374528379           eng        796         191531   
1408  0140449248  9780140449242           eng       1013           1673   
1409  0679729259  9780679729259           eng        796            617   
1410  1596440791  9781596440791           eng         16             20   
1411  159308045x  9781593080457           eng        720           1089   
1634  0099922800  9780099922803           eng        796            443   
1990  0553212168  9780553212167           eng       1072           1022   
9356  1596440783  9781596440784           eng         16              3   

      text_reviews_count  ... edition_avgRating added_toShelves  \
1406                  91  ...              4.27          1638.0   
1407                6795  ...              4.35        855582.0   
1408                 184  ...              4.44         11774.0   
1409                  80  ...              4.51          1381.0   
1410                   2  ...              4.52           164.0   
1411                 202  ...              4.38          3421.0   
1634                  55  ...              4.45          1657.0   
1990                 154  ...              4.37          3236.0   
9356                   1  ...              4.67             9.0   

     added_toShelve text_reviews_coun        first_author  num_contributors  \
1406            NaN               NaN  fyodor dostoyevsky                 4   
1407            NaN               NaN  fyodor dostoyevsky                 4   
1408            NaN               NaN  fyodor dostoyevsky                 2   
1409            NaN               NaN  fyodor dostoyevsky                 3   
1410            NaN               NaN  fyodor dostoyevsky                 3   
1411            NaN               NaN  fyodor dostoyevsky                 3   
1634            NaN               NaN  fyodor dostoyevsky                 3   
1990            NaN               NaN  fyodor dostoyevsky                 3   
9356            NaN               NaN  fyodor dostoyevsky                 3   

      size_of_publisher  num_book_per_author  is_english book_count  
1406                  3                   37           1          9  
1407                  3                   37           1          9  
1408                  3                   37           1          9  
1409                  3                   37           1          9  
1410                  2                   37           1          9  
1411                  3                   37           1          9  
1634                  3                   37           1          9  
1990                  3                   37           1          9  
9356       

### Create a "is_serie" column

In [497]:
# Create a regular expression pattern to match titles containing numbers
pattern = r'\b\d+\b'  # This pattern matches one or more digits

# Use the str.contains() method with the regex pattern to filter the DataFrame
books_with_number = df_scraped[df_scraped['title'].str.contains(pattern)]

# Print or further process the extracted books
books_with_number['title'].sample(20)

7014             the final battle tales from the odyssey 6
3815                               tehanu earthsea cycle 4
6401                            casino royale james bond 1
5907                the princess bride long tall texans 15
705      murder at the washington tribune capital crime...
10474                 nimitz class admiral arnold morgan 1
7263     algorithms in c parts 14 fundamentals data str...
9192                on basilisk station honor harrington 1
5701                               already dead joe pitt 1
3137                    1 000 places to see before you die
8923               lords and ladies discworld 14 witches 4
963                               greek tragedies volume 2
7006        night of blood dragonlance the minotaur wars 1
8310                   astonishing xmen volume 2 dangerous
827                             the lost boy dave pelzer 2
11063                          schmidt delivered schmidt 2
6079                                                犬夜叉 

In [498]:
# Add a new column 'is_serie' with 1 for books with a number in the title and 0 for books without numbers
df_scraped['is_serie'] = np.where(df_scraped['title'].str.contains(pattern), 1, 0)

In [499]:
df_scraped.sample(10)

bookID                                              title  \
6753    25347                     out of egypt christ the lord 1   
5261    18981        hayao miyazaki master of japanese animation   
3620    13113  zen and the art of motorcycle maintenance an i...   
1115     3768                                   book of sketches   
8497    32627                        everville book of the art 2   
10002   40289                                          the abyss   
9727    38830                             stranger from the past   
3930    13996                             alta dragon jousters 2   
8492    32592                                      in wonderland   
7023    26653                                    the coal tattoo   

                                      authors  average_rating        isbn  \
6753                                anne rice            3.58  0345436830   
5261                           helen mccarthy            3.93  1880656418   
3620                         robert m. pirsig            3.77  0688052304   
1115                jack kerouac/george condo            3.80  0142002151   
8497                             clive barker            4.03  0060933151   
10002  orson scott card/james francis cameron            4.07  0099690608   
9727                             penny jordan            3.23  0373115997   
3930                          mercedes lackey            4.03  0756402573   
8492              knut hamsun/sverre lyngstad            3.60  0970312555   
7023                              silas house            4.04  0345480058   

              isbn13 language_code  num_pages  ratings_count  \
6753   9780345436832           eng        350          11519   
5261   9781880656419         en-us        240            317   
3620   9780688052300           eng        412            424   
1115   9780142002155           eng        432            572   
8497   9780060933159           eng        697           8829   
10002  9780099690603           eng        363          17911   
9727   9780373115990           eng        224             29   
3930   9780756402570         en-us        434           7060   
8492   9780970312556           nor        185             86   
7023   9780345480057           eng        368           1789   

       text_reviews_count  ... added_toShelves added_toShelve  \
6753                  981  ...         21868.0            NaN   
5261                   19  ...          1199.0            NaN   
3620                   47  ...          1155.0            NaN   
1115                   43  ...          1576.0            NaN   
8497                  157  ...         17013.0            NaN   
10002                 111  ...         30412.0            NaN   
9727                    7  ...            73.0            NaN   
3930                  113  ...         11224.0            NaN   
8492                    8  ...           342.0            NaN   
7023                  144  ...          3855.0            NaN   

      text_reviews_coun      first_author num_contributors  size_of_publisher  \
6753                NaN         anne rice                1                  3   
5261                NaN    helen mccarthy                1                  2   
3620                NaN   robert m pirsig                1                  2   
1115                NaN      jack kerouac                2                  3   
8497                NaN      clive barker                1                  2   
10002               NaN  orson scott card                2                  1   
9727                NaN      penny jordan                1                  2   
3930                NaN   mercedes lackey                1                  3   
8492                NaN       knut hamsun                2                  1   
7023                NaN       silas house                1                  3   

       num_book_per_author  is_english  book_count is_serie  
6753                    23         

In [500]:
#Let's see if we have to add some books in "is_serie" based on word in there title
df_scraped[(df_scraped['is_serie'] == '0') & (df_scraped['title'].str.contains('trilogy|tome|chronicles|series', case=False))]

Empty DataFrame
Columns: [bookID, title, authors, average_rating, isbn, isbn13, language_code, num_pages, ratings_count, text_reviews_count, publication_date, publisher, first_published, book_format, new_publisher, edition_avgRating, added_toShelves, added_toShelve, text_reviews_coun, first_author, num_contributors, size_of_publisher, num_book_per_author, is_english, book_count, is_serie]
Index: []

[0 rows x 26 columns]

### Convert the first_published date into 3 columns : published_year, published_month, published_day

By looking at the data, we notice that for some line, the first_published date didn't extracted well, and it's written "ary" instead of january or february.

In [501]:
df_scraped[['title','first_published']][df_scraped['first_published'].str.contains(r'ary', case=False) & ~df_scraped['first_published'].str.contains(r'january|february', case=False)]

title first_published
56                       simply beautiful beaded jewelry    ary 28, 2006
77     the power of one the solo play for playwrights...     ary 7, 2000
124                                  tropic of capricorn     ary 1, 1939
176                                     gravitys rainbow    ary 28, 1973
179                                     gravitys rainbow    ary 28, 1973
...                                                  ...             ...
11067                                     the home front    ary 23, 1989
11090                 la conspiración de los alquimistas     ary 1, 1999
11092                   the call of the mall how we shop     ary 2, 2004
11101  undaunted courage the pioneering first mission...    ary 15, 1996
11102                                  undaunted courage    ary 15, 1996

[358 rows x 2 columns]

In [502]:
#We will replace the "ary" in those line by "february" 
# Filter rows containing "ary" but not "january" or "february"
filtered_rows = df_scraped['first_published'].str.contains(r'ary', case=False) & ~df_scraped['first_published'].str.contains(r'january|february', case=False)

# Replace the filtered results with "february"
df_scraped.loc[filtered_rows, 'first_published'] = df_scraped.loc[filtered_rows, 'first_published'].str.replace(r'ary', 'february', case=False)

In [503]:
df_scraped[['title','first_published']][df_scraped['first_published'].str.contains('published', case=False)]

title  \
133                                         love letters   
201                       timbuktu leviathan moon palace   
203    the coming economic collapse how you can thriv...   
229    guidebook to zen and the art of motorcycle mai...   
263                 best of london lonely planet best of   
...                                                  ...   
10881              the best american travel writing 2006   
10898  poems between women four centuries of love rom...   
10937  un amour de swann à la recherche du temps perd...   
10947  poetry and prose of alexander pope riverside e...   
11106                                   fantastic mr fox   

                   first_published  
133      published january 1, 1999  
201     published october 31, 2002  
203    published february 21, 2006  
229    published november 19, 1990  
263      published january 1, 2004  
...                            ...  
10881   published october 11, 2006  
10898     published april 15, 1999  
10937    published january 1, 2006  
10947    published january 2, 1968  
11106      published july 21, 2003  

[160 rows x 2 columns]

In [504]:
#Let's delete the word "published" from the first_published column
df_scraped['first_published'] = df_scraped['first_published'].str.replace('published ', '')

By trying to convert the "first_published" column into date, we notice several errors due to date before 1677. Indeed, the smallest date that pandas can handle is January 1, 1677. As we are looking for an average rating given on a website, there is probably not a big difference between books written in 1524 or 1678, as they are considered "old". By lack of time, we'll change all year below 1677 by 1678.

In [505]:
# The smallest date that pandas can handle is January 1, 1677. Let's remove all date below 1677 and change the year by 1678
for i, date_str in enumerate(df_scraped['first_published']):
    # Use regular expression to find the year in the string
    match = re.search(r'\b\d{1,4}\b$', date_str)
    if match:
        year = int(match.group())
        if year < 1677:
            # Replace the year with 1678
            modified_date_str = re.sub(r'\b\d{1,4}\b$', '1678', date_str)
            df_scraped.at[i, 'first_published'] = modified_date_str

In [506]:
df_scraped['first_published'] = pd.to_datetime(df_scraped['first_published'], format='%B %d, %Y')

In [507]:
# Convert the first_published date into 3 columns : published_year, published_month, published_day
df_scraped['published_year'] = df_scraped['first_published'].dt.year
df_scraped['published_month'] = df_scraped['first_published'].dt.month
df_scraped['published_day'] = df_scraped['first_published'].dt.day


In [508]:
df_scraped.tail(20)

bookID                                              title  \
11107   45572                    fantastic mr fox cover to cover   
11108   45574                                     ajax in action   
11109   45583                      historia del rey transparente   
11110   45585                                     bella y oscura   
11111   45592                       la tía julia y el escribidor   
11112   45595                       la tía julia y el escribidor   
11113   45604                        narraciones extraordinarias   
11114   45607                             las crónicas de narnia   
11115   45615           o trono de prata as crónicas de nárnia 6   
11116   45616           a última batalha as crónicas de nárnia 7   
11117   45617     o cavalo e o seu rapaz as crónicas de nárnia 3   
11118   45623       o sobrinho do mágico as crónicas de nárnia 1   
11119   45625  a viagem do caminheiro da alvorada as crónicas...   
11120   45626         o príncipe caspian as crónicas de nárnia 4   
11121   45630                                  whores for gloria   
11122   45631     expelled from eden a william t vollmann reader   
11123   45633                        you bright and risen angels   
11124   45634                        the iceshirt seven dreams 1   
11125   45639                                        poor people   
11126   45641                        las aventuras de tom sawyer   

                                                 authors  average_rating  \
11107                                         roald dahl            4.05   
11108            dave crane/eric pascarello/darren james            3.56   
11109                                       rosa montero            3.90   
11110                                       rosa montero            3.66   
11111                                 mario vargas llosa            3.92   
11112                                 mario vargas llosa            3.92   
11113                                    edgar allan poe            4.13   
11114  c.s. lewis/margarita e. valdes/gemma gallart/p...            4.26   
11115                       c.s. lewis/ana falcão bastos            3.96   
11116        c.s. lewis/pauline baynes/ana falcão bastos            4.03   
11117        c.s. lewis/pauline baynes/ana falcão bastos            3.92   
11118        c.s. lewis/pauline baynes/ana falcão bastos            4.04   
11119        c.s. lewis/pauline baynes/ana falcão bastos            4.09   
11120        c.s. lewis/pauline baynes/ana falcão bastos            3.97   
11121                                william t. vollmann            3.69   
11122  william t. vollmann/larry mccaffery/michael he...            4.06   
11123                                william t. vollmann            4.08   
11124                                william t. vollmann            3.96   
11125                                william t. vollmann            3.72   
11126                                         mark twain            3.91   

             isbn         isbn13 language_code  num_pages  ratings_count  \
11107  1855495090  9781855495098           eng         90              7   
11108  1932394613  9781932394610         en-us        680            126   
11109  8466318771  9788466318778           spa        592           1266   
11110  843221728x  9788432217289           spa        204            293   
11111  8466302298  9788466302296           spa        566            162   
11112  8432203238  9788432203237           spa        447             25   
11113  9583006408  9789583006401           spa        316             36   
11114  0061199001  9780061199004           spa        816            186   
11115  972233168x  9789722331685           por        168            141   
11116  9722332201  9789722332200           por        149            211   
11117  9722330551  9789722330558           por        160            207   
11118  9722329987  9789722329989           por        147            